<a href="https://colab.research.google.com/github/vishwajeet1132/Collab/blob/main/Solution_Artificial_Neural_Network_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import Library
import pandas as pd

In [2]:
#Import the files to Google Colab
url = 'https://raw.githubusercontent.com/rc-dbe/bigdatacertification/master/dataset/churn_trasnsformed_new.csv'
df_csv = pd.read_csv(url, sep=',',)

# Show 10 first Row
df_csv.head()

Unnamed: 0  gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  \
0           0         0         1                0                1   
1           1         1         0                0                1   
2           2         1         0                0                1   
3           3         1         0                0                1   
4           4         0         1                0                1   

   Partner_0  Partner_1  Dependents_0  Dependents_1  tenure  ...  Contract_1  \
0          0          1             0             1       1  ...           0   
1          1          0             0             1      34  ...           1   
2          1          0             0             1       2  ...           0   
3          1          0             0             1      45  ...           1   
4          1          0             0             1       2  ...           0   

   Contract_2  PaperlessBilling_0  PaperlessBilling_1  PaymentMethod_0  \
0           1                   0                   1                0   
1           0                   1                   0                0   
2           1                   0                   1                0   
3           0                   1                   0                0   
4           1                   0                   1                0   

   PaymentMethod_1  PaymentMethod_2  MonthlyCharges  TotalCharges  Churn  
0                0                1            2985        2985.0      0  
1                1                0            5695       18895.0      0  
2                1                0            5385       10815.0      1  
3                1                1             423      184075.0      0  
4                0                1             707       15165.0      1  

[5 rows x 47 columns]

In [3]:
# Remove "Unnamed:O" Coloumn
df = df_csv.drop("Unnamed: 0", axis=1)
df.head()

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  Partner_1  \
0         0         1                0                1          0          1   
1         1         0                0                1          1          0   
2         1         0                0                1          1          0   
3         1         0                0                1          1          0   
4         0         1                0                1          1          0   

   Dependents_0  Dependents_1  tenure  PhoneService_0  ...  Contract_1  \
0             0             1       1               0  ...           0   
1             0             1      34               1  ...           1   
2             0             1       2               1  ...           0   
3             0             1      45               0  ...           1   
4             0             1       2               1  ...           0   

   Contract_2  PaperlessBilling_0  PaperlessBilling_1  PaymentMethod_0  \
0           1                   0                   1                0   
1           0                   1                   0                0   
2           1                   0                   1                0   
3           0                   1                   0                0   
4           1                   0                   1                0   

   PaymentMethod_1  PaymentMethod_2  MonthlyCharges  TotalCharges  Churn  
0                0                1            2985        2985.0      0  
1                1                0            5695       18895.0      0  
2                1                0            5385       10815.0      1  
3                1                1             423      184075.0      0  
4                0                1             707       15165.0      1  

[5 rows x 46 columns]

In [4]:
# Check the Data Infomation
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 46 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   gender_0            7043 non-null   int64  
 1   gender_1            7043 non-null   int64  
 2   SeniorCitizen_0     7043 non-null   int64  
 3   SeniorCitizen_1     7043 non-null   int64  
 4   Partner_0           7043 non-null   int64  
 5   Partner_1           7043 non-null   int64  
 6   Dependents_0        7043 non-null   int64  
 7   Dependents_1        7043 non-null   int64  
 8   tenure              7043 non-null   int64  
 9   PhoneService_0      7043 non-null   int64  
 10  PhoneService_1      7043 non-null   int64  
 11  MultipleLines_0     7043 non-null   int64  
 12  MultipleLines_1     7043 non-null   int64  
 13  MultipleLines_2     7043 non-null   int64  
 14  InternetService_0   7043 non-null   int64  
 15  InternetService_1   7043 non-null   int64  
 16  Intern

In [5]:
#Import MinMax Scaler
from sklearn.preprocessing import MinMaxScaler

# initialize min-max scaler
mm_scaler = MinMaxScaler()
column_names = df.columns.tolist()
column_names.remove('Churn')

# Transform all attributes
df[column_names] = mm_scaler.fit_transform(df[column_names])
df.sort_index(inplace=True)
df.head()

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  Partner_1  \
0       0.0       1.0              0.0              1.0        0.0        1.0   
1       1.0       0.0              0.0              1.0        1.0        0.0   
2       1.0       0.0              0.0              1.0        1.0        0.0   
3       1.0       0.0              0.0              1.0        1.0        0.0   
4       0.0       1.0              0.0              1.0        1.0        0.0   

   Dependents_0  Dependents_1    tenure  PhoneService_0  ...  Contract_1  \
0           0.0           1.0  0.013889             0.0  ...         0.0   
1           0.0           1.0  0.472222             1.0  ...         1.0   
2           0.0           1.0  0.027778             1.0  ...         0.0   
3           0.0           1.0  0.625000             0.0  ...         1.0   
4           0.0           1.0  0.027778             1.0  ...         0.0   

   Contract_2  PaperlessBilling_0  PaperlessBilling_1  PaymentMethod_0  \
0         1.0                 0.0                 1.0              0.0   
1         0.0                 1.0                 0.0              0.0   
2         1.0                 0.0                 1.0              0.0   
3         0.0                 1.0                 0.0              0.0   
4         1.0                 0.0                 1.0              0.0   

   PaymentMethod_1  PaymentMethod_2  MonthlyCharges  TotalCharges  Churn  
0              0.0              1.0        0.250169      0.003420      0  
1              1.0              0.0        0.478745      0.021766      0  
2              1.0              0.0        0.452598      0.012449      1  
3              1.0              1.0        0.034076      0.212235      0  
4              0.0              1.0        0.058030      0.017465      1  

[5 rows x 46 columns]

In [6]:
# Selecting the Feature, by remove the unused feature
feature = ['Churn']
train_feature = df.drop(feature, axis=1)

# Set The Target
train_target = df["Churn"]

In [7]:
# Show the Feature
train_feature.head(5)

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  Partner_1  \
0       0.0       1.0              0.0              1.0        0.0        1.0   
1       1.0       0.0              0.0              1.0        1.0        0.0   
2       1.0       0.0              0.0              1.0        1.0        0.0   
3       1.0       0.0              0.0              1.0        1.0        0.0   
4       0.0       1.0              0.0              1.0        1.0        0.0   

   Dependents_0  Dependents_1    tenure  PhoneService_0  ...  Contract_0  \
0           0.0           1.0  0.013889             0.0  ...         0.0   
1           0.0           1.0  0.472222             1.0  ...         0.0   
2           0.0           1.0  0.027778             1.0  ...         0.0   
3           0.0           1.0  0.625000             0.0  ...         0.0   
4           0.0           1.0  0.027778             1.0  ...         0.0   

   Contract_1  Contract_2  PaperlessBilling_0  PaperlessBilling_1  \
0         0.0         1.0                 0.0                 1.0   
1         1.0         0.0                 1.0                 0.0   
2         0.0         1.0                 0.0                 1.0   
3         1.0         0.0                 1.0                 0.0   
4         0.0         1.0                 0.0                 1.0   

   PaymentMethod_0  PaymentMethod_1  PaymentMethod_2  MonthlyCharges  \
0              0.0              0.0              1.0        0.250169   
1              0.0              1.0              0.0        0.478745   
2              0.0              1.0              0.0        0.452598   
3              0.0              1.0              1.0        0.034076   
4              0.0              0.0              1.0        0.058030   

   TotalCharges  
0      0.003420  
1      0.021766  
2      0.012449  
3      0.212235  
4      0.017465  

[5 rows x 45 columns]

In [8]:
# Split Data
from sklearn.model_selection import train_test_split, cross_val_score
X_train, X_test, y_train, y_test = train_test_split(train_feature ,train_target, shuffle = True, test_size=0.9, random_state=1)

In [9]:
# Show the training data
X_train.head()

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  \
4342       0.0       1.0              1.0              0.0        1.0   
775        0.0       1.0              0.0              1.0        0.0   
4871       1.0       0.0              0.0              1.0        0.0   
4648       1.0       0.0              0.0              1.0        1.0   
5209       1.0       0.0              0.0              1.0        0.0   

      Partner_1  Dependents_0  Dependents_1    tenure  PhoneService_0  ...  \
4342        0.0           0.0           1.0  0.722222             1.0  ...   
775         1.0           0.0           1.0  1.000000             1.0  ...   
4871        1.0           1.0           0.0  0.805556             1.0  ...   
4648        0.0           0.0           1.0  0.027778             1.0  ...   
5209        1.0           0.0           1.0  1.000000             1.0  ...   

      Contract_0  Contract_1  Contract_2  PaperlessBilling_0  \
4342         0.0         1.0         1.0                 1.0   
775          0.0         1.0         1.0                 0.0   
4871         0.0         1.0         0.0                 0.0   
4648         0.0         0.0         1.0                 0.0   
5209         0.0         1.0         1.0                 0.0   

      PaperlessBilling_1  PaymentMethod_0  PaymentMethod_1  PaymentMethod_2  \
4342                 0.0              0.0              1.0              1.0   
775                  1.0              1.0              0.0              0.0   
4871                 1.0              1.0              0.0              0.0   
4648                 1.0              0.0              0.0              1.0   
5209                 1.0              1.0              0.0              0.0   

      MonthlyCharges  TotalCharges  
4342        0.503205      0.357814  
775         0.931680      0.930030  
4871        0.061825      0.049571  
4648        0.057692      0.000144  
5209        0.925776      0.090524  

[5 rows x 45 columns]

To train the ANN Model.We will use the MLPClassifier from Scikit Learn Library. The full documentation can be seen [HERE](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html). Below is the default parameter:

`sklearn.neural_network.MLPClassifier(hidden_layer_sizes=(100), activation=’relu’, solver=’adam’, alpha=0.0001, batch_size=’auto’, learning_rate=’constant’, learning_rate_init=0.001, power_t=0.5, max_iter=200, shuffle=True, random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10)`

In [10]:
# Import Library
from sklearn.neural_network import MLPClassifier

# Fitting Model
mlp = MLPClassifier(hidden_layer_sizes=(5), activation = 'relu', solver = 'adam',max_iter= 10000, verbose = True)
mlp = mlp.fit(X_train,y_train)

# Prediction to Test Dataset
y_predmlp = mlp.predict(X_test)

Iteration 1, loss = 1.81038403
Iteration 2, loss = 1.68557175
Iteration 3, loss = 1.57081771
Iteration 4, loss = 1.46788710
Iteration 5, loss = 1.37683216
Iteration 6, loss = 1.29812450
Iteration 7, loss = 1.23207852
Iteration 8, loss = 1.17884706
Iteration 9, loss = 1.13256023
Iteration 10, loss = 1.09594324
Iteration 11, loss = 1.06558449
Iteration 12, loss = 1.04197361
Iteration 13, loss = 1.02265871
Iteration 14, loss = 1.00715309
Iteration 15, loss = 0.99442992
Iteration 16, loss = 0.98353528
Iteration 17, loss = 0.97494123
Iteration 18, loss = 0.96694410
Iteration 19, loss = 0.96031478
Iteration 20, loss = 0.95456264
Iteration 21, loss = 0.94943969
Iteration 22, loss = 0.94464854
Iteration 23, loss = 0.94053559
Iteration 24, loss = 0.93662121
Iteration 25, loss = 0.93304834
Iteration 26, loss = 0.92974181
Iteration 27, loss = 0.92673172
Iteration 28, loss = 0.92382804
Iteration 29, loss = 0.92113332
Iteration 30, loss = 0.91843669
Iteration 31, loss = 0.91589157
Iteration 32, los

In [11]:
print('Number of Layer =', mlp.n_layers_)
print('Number of Iteration =', mlp.n_iter_)
print('Current loss computed with the loss function =', mlp.loss_)

Number of Layer = 3
Number of Iteration = 1633
Current loss computed with the loss function = 0.37002628785666014


Since it was the classification problem, we can evaluate the model using Confussion Matrix

In [12]:
# Import the metrics class
from sklearn import metrics


# Confussion Matrix
cnf_matrixmlp = metrics.confusion_matrix(y_test, y_predmlp)
cnf_matrixmlp

array([[4013,  661],
       [ 795,  870]])

In [13]:
# Show the Accuracy, Precision, Recall, F1, etc.
acc_mlp = metrics.accuracy_score(y_test, y_predmlp)
prec_mlp = metrics.precision_score(y_test, y_predmlp)
rec_mlp = metrics.recall_score(y_test, y_predmlp)
f1_mlp = metrics.f1_score(y_test, y_predmlp)
kappa_mlp = metrics.cohen_kappa_score(y_test, y_predmlp)

print("Accuracy:", acc_mlp)
print("Precision:", prec_mlp)
print("Recall:", rec_mlp)
print("F1 Score:", f1_mlp)
print("Cohens Kappa Score:", kappa_mlp)

Accuracy: 0.7703107745701214
Precision: 0.5682560418027433
Recall: 0.5225225225225225
F1 Score: 0.5444305381727158
Cohens Kappa Score: 0.39123713971717566


## Split Data cell Explaination
X_train, X_test, y_train, y_test = train_test_split(train_feature ,train_target, shuffle = True, test_size=0.3, random_state=1)

In the above code we used **train_test_split** to divide the data in two parts

1) Test Data

2) Train Data

The func **test_train_split** takes input param test_size (0.3 in this case) which tells how much data of dataframe to be considered as test_data (0.3 means 30% of the dataframe)

Train data: The data which will be used to Train the Model (MLPClassifier) in this case.

Test data: The data which will be used to check the accuracy of the trained model.


## Confussion Matrix cell Explaination
To test the how effective our model is we use confusion Matrix.

Confusion matrix is a used to measure the performance of our trained Model
It provides us with four values

                                [True Positive ,False Positive

                                 False Negative , True Negative]



True Positive: Prediction is positive and it is True (TP)

True Negative: Prediction is negative and it True (TN)

False Positive: Prediction is Positive and it is False (FP)

False Negative: Prediction is Negative and it is False (FN)



## Show the Accuracy, Precision, Recall, F1
Accuracy:
Accuracy measures the correctness of your model's predictions. It calculates the ratio of correctly predicted instances to the total number of instances in the dataset.

Accuracy = (TP + TN) / (TP + TN + FP + FN)

Precision:
Precision measures the ratio of correctly predicted positive instances (True Positives, TP) to all instances predicted as positive (True Positives + False Positives, TP + FP). It focuses on the accuracy of positive predictions.

Precision = TP / (TP + FP)

Recall (Sensitivity or True Positive Rate):
Recall measures the ratio of correctly predicted positive instances (True Positives, TP) to all actual positive instances (True Positives + False Negatives, TP + FN). It quantifies how many of the actual positive instances were correctly predicted.

Recall = TP / (TP + FN)

F1 Score:
The F1 Score is the harmonic mean of precision and recall. It balances precision and recall and is particularly useful when dealing with imbalanced datasets, where one class has significantly more samples than the other.

F1 Score = 2 * (Precision * Recall) / (Precision + Recall)



## Results with 70% train data and 30% test data
confuison matrix
```
array  ([[1400,  185],
         [ 223,  305]])
```
Accuracy: 0.8069096071935636

Precision: 0.6224489795918368

Recall: 0.5776515151515151

F1 Score: 0.5992141453831041

Cohens Kappa Score: 0.47226544661647885



## Results with 10% train data and 90% test data
confuison matrix
```
array([[4013,  661],
       [ 795,  870]])
```
Accuracy: 0.7703107745701214

Precision: 0.5682560418027433

Recall: 0.5225225225225225

F1 Score: 0.5444305381727158

Cohens Kappa Score: 0.39123713971717566



## Comparing the performance
As we can see By changing the test train split, all the Performance Metrices went down but not too much
maybe

1) Because there is no much deviation in overall data

2) The data is not sufficient to train a model


In [ ]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.multioutput import MultiOutputRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
url = 'https://raw.githubusercontent.com/rc-dbe/bigdatacertification/master/dataset/churn_trasnsformed_new.csv'
df_csv = pd.read_csv(url, sep=',',)
df = df_csv.drop("Unnamed: 0", axis=1)
mm_scaler = MinMaxScaler()
column_names = df.columns.tolist()
df[column_names] = mm_scaler.fit_transform(df[column_names])
df.sort_index(inplace=True)
feature_cols = [col for col in df.columns if col != 'TotalCharges']
train_feature = df[feature_cols]
train_target = df[['TotalCharges']]
X_train, X_test, y_train, y_test = train_test_split(train_feature ,train_target, shuffle = True, test_size=0.3, random_state=100)
base_regressor = RandomForestRegressor()
multi_output_regressor = MultiOutputRegressor(base_regressor)
multi_output_regressor.fit(X_train, y_train)
y_pred = multi_output_regressor.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")

Mean Squared Error: 0.031474449606855526
